spec/controllers/admin/subjects_controller_spec.rb

```diff
@@ -0,0 +1,175 @@
+require "rails_helper"
+include SessionsHelper
+
+RSpec.describe Admin::SubjectsController, type: :controller do
+  let!(:user) {FactoryBot.create :user, role: :user}
+  let!(:admin) {FactoryBot.create :user}
+  let(:valid_params) {{subject: {name: "Bài thi sun", content: "Đề thi về ruby sun"}}}
```

Comment:
let và let! Khác j nhau cu

Reply:
let chỉ có giá trị trong từng example (trong Rspec thì được đóng gói trong khối it), và nếu như bạn gọi đến một biến let nhiều lần trong cùng một example thì nó sẽ chỉ khai báo khối let một lần ở lần gọi đầu tiên.
let! được gọi trong một khối before ngầm, do đó kết quả đã được tính toán và cache lại trước khi khổi it được thực thi.